CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

IntegralProgrammer opened the pull request:
This PR implements CARDS-1759 by providing a data export task that can either be manually triggered or periodically scheduled. This data export task begins by exporting a list of all _Form_ nodes and a list of all _Subject_ nodes that are present in the JCR at the time of execution of the data export task thus giving a snapshot of which _Forms_ and _Subjects_ are available in CARDS at a moment in time. The data export task then serializes to JSON all _Form_ and _Subject_ nodes that were last-modified within a given time window thus allowing for more efficient data backups where only modified data is exported as opposed to redundantly exporting already-exported data.

To restore data, a data restoration script is given a list of _Subject_ nodes to create, a list of _Form_ nodes to create, and a directory where the JSON serialized _Subject_ and _Form_ nodes can be found.

To test this Pull Request:

- CARDS must be started with the `BACKUP_WEBHOOK_URL` environment variable set to `http://localhost:8012`.
- Start the `backup_recorder.js` server by entering the `Utilities/Administration/Backup-To-JSON` directory and running the command `nodejs backup_recorder.js /path/to/backup/directory`.
- Backups can be initiated by visiting `http://localhost:8080/`, logging in as `admin`, and then visiting `http://localhost:8080/Subjects.webhookbackup?dateLowerBound=YYYY-MM-DDThh:mm:ss` or `http://localhost:8080/Subjects.webhookbackup?dateLowerBound=YYYY-MM-DDThh:mm:ss&dateUpperBound=YYYY-MM-DDThh:mm:ss`, replacing `YYYY-MM-DDThh:mm:ss` with a date/time for a lower or upper date/time bound (eg. `2022-05-18T12:00:00` for 12:00:00 PM on May 18th 2022).
- Restoring data to a fresh CARDS instance should be done with the CARDS instance started with the `COMPUTED_ANSWERS_DISABLED` environment variable set to `true`. To ensure that you are starting a fresh CARDS instance, run `rm -rf .cards-data` before starting the new instance.
- To restore data, run the `restore_json_backup.py` script as `python3 restore_json_backup.py /path/to/backup/directory /path/to/backup/directory/FormListBackup_TIMESTAMP.json /path/to/backup/directory/SubjectListBackup_TIMESTAMP.json`.

With these steps in mind, please proceed into testing that the Forms/Subjects from the following CARDS projects can be successfully backed-up and restored:

- Cards4Proms (DATA-PRO)
- Cards4LFS
- Cards4CaRe
- Cards4Kids (WilliamsDB)

Also verify that:
- _Note_ sections are properly backed-up and restored.
- Pedigree answers are properly backed-up and restored (create a Pedigree chart in the LFS _Patient information_ Form)
- File answers are properly backed-up and restored (you can use the LFS _Somatic Variants_ Form for this)
- Slack notifications for backup tasks are working properly. You can start the mock Slack server by entering the `Utilities/Development` directory and running the command `nodejs mock_slack.js`. CARDS should be started with the `SLACK_BACKUP_NOTIFICATIONS_URL` set to `http://localhost:8000/`.
- Nightly backups can be scheduled to run automatically through the `NIGHTLY_WEBHOOK_BACKUP_SCHEDULE` environment variable.

To obtain estimates of how large the backups will be for Cards4Proms (DATA-PRO) backups:

- Build the `CARDS-1759_size-measurements` branch
- Start the Mock Torch server and instruct it to serve 100 randomly generated patients
```bash
cd Utilities/Development
nodejs mock_torch.js --appointment-time-hours-from-now=24 --nRandomPatients=100
```
- Start CARDS in Cards4Proms mode with the `BACKUP_WEBHOOK_URL` environment variable set to `http://localhost:8012`.
- Visit `http://localhost:8080/system/console/configMgr` and configure _PROMs import_ as follows:
  - `Name`: `LocalNodeJS`
  - `Import schedule`: `0 0 0 * * ? *`
  - `days to query`: `3`
  - `Endpoint URL`: `http://localhost:8011/`
  - `Authentication URL`: (leave blank)
  - `Vault token`: (leave blank)
  - `Clinic names`: `6012-HC-Congenital Cardiac`
  - `Provider names`: (leave blank)
  - `Allowed provider roles`: (leave blank)
  - `Vault role name`: (leave blank)
  - `Dates to query`: (leave blank)
 - Import the randomly generated patients/visits by pointing your browser to `http://localhost:8080/Subjects.importTorch?config=LocalNodeJS`
 - Visit the main dashboard at `http://localhost:8080` and keep periodically refreshing the page until all 100 patients and visits have been loaded.
 - Load randomly generated survey responses:
```bash
cd Utilities/Development
python3 auto_answer_surveys.py Cards4Proms-SampleAnswers/AUDITC.csv Cards4Proms-SampleAnswers/EQ5D.csv Cards4Proms-SampleAnswers/GAD7.csv Cards4Proms-SampleAnswers/PHQ9.csv Cards4Proms-SampleAnswers/SC.csv
```
- Follow the instructions in `Utilities/Utilities/Administration/Backup-To-JSON/README.md` to perform a complete backup to JSON of this data followed by a restore of this data into a blank Cards4Proms instance. Ensure that the restored data matches the backed up data.

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Bring in the NightlyExportTask code from the Cards4CaRe project as the webhook-backup CARDS module

Right now, the WebhookBackupTask can only be invoked through the /Subjects.webhookbackup endpoint as the quartz scheduler has not yet been configured for it.

A debugging call to LOGGER.warn has been added to the start of the WebhookBackupTask::output which displays the data that is to be exported to the webhook backup service

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Send data as the body of a POST request instead of a push to an S3 bucket

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Code clean up

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Remove "bare" and "relativeDates" filters

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Convert dateLowerBound and dateUpperBound URL parameters from LocalDate typed to LocalDateTime typed

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Code cleanup

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

During backup, push the present list of Form and Subject names to the backup server so that during restore, Forms and Subjects that were deleted before backup will not be un-deleted during restore

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Working implementation of backup to directory containing files; FormListBackup and SubjectListBackup and directories; blobs, Forms, Subjects

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Code cleanup

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Code cleanup

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Code cleanup

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Remove unnecessary code from WebhookBackupTask.java

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

If an upper bound for jcr:lastModified datetime is not given, assume it to be positive infinity

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Refactor: rename strToDate to strToDateTime

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Code cleanup

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Add scheduler to call WebhookBackupTask::doNightlyExport as dictated by the NIGHTLY_WEBHOOK_BACKUP_SCHEDULE environment variable containing a quartz cron string

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Implemented WebhookBackupTask.doNightlyExport

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Log when the webhook backup starts, finishes, and fails

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

If the backup server response with a non 2xx HTTP status, consider the backup to have failed

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Notify Slack of the started/finished/failed backups

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Refactor: Create the DATE_TIME_JCR_FORMAT String constant

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

If SLACK_NOTIFICATIONS_URL environment variable is not defined, simply skip sending notifications to Slack instead of stopping the entire backup process

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Add the Subject's "identifier" to the serialized Subject's JSON file

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

restore_json_backup.py now restores Subjects into the CARDS JCR

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

restore_json_backup.py script now restores Forms, Answers, and AnswerSections to the CARDS JCR

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Add the ability to disable computation of computed answers by specifying the environment variable COMPUTED_ANSWERS_DISABLED=true

This is necessary to prevent duplicated computed answers upon data restoration

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

backup_recorder.js now saves backup JSON files to a directory specified by a CLI argument

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Move backup_recorder.js and restore_json_backup.py to the Utilities/Administration/Backup-To-JSON directory

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

backup_recorder.js should only listen on localhost (127.0.0.1) and not all interfaces (0.0.0.0)

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Add a README.md file describing how to backup/restore Forms and Subjects to/from JSON files

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Fix response message to /Subjects.webhookbackup - we're not necessarily using an S3 bucket

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Move backup_recorder.js to port 8012 so that it can be run concurrently with mock_slack.js and mock_torch.js

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Add backup/restore of answer notes

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Add security warning to Utilities/Administration/Backup-To-JSON/README.md

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

backup_recorder.js now supports saving FileAnswers with multiple uploaded files. Uploaded names are now also preserved

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

restore_json_backup.py now restores files belonging to cards:FileAnswer nodes

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Minor refactoring and adding support for "image" property used in cards:PedigreeAnswer nodes

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

In restore_json_backup.py, properly handle the situation where a JSON-serialized answer node does not have a "value" property

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

In restore_json_backup.py, properly handle the situation where a listed Subject or Form does not have an associated JSON file. Also handle the situation where a Form is created with a reference to a Subject that does not exist

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Remove old comment, cards:FileAnswer answers can now be properly restored

sdumitriu commented on Utilities/Administration/Backup-To-JSON/backup_recorder.js:
```diff
@@ -0,0 +1,262 @@
+  return h.digest().toString('hex');
+};
+
+const getFileDataBuffer = (jsonValue) => {
+  let bytesArray = [];
+  for (let i = 0; i < jsonValue.length; i++) {
+    bytesArray.push(jsonValue.charCodeAt(i));
+  }
+  return Buffer.from(bytesArray);
+};
```
Why not `Buffer.from(jsonValue, 'utf-8')`?

IntegralProgrammer replied:
I tried it but it doesn't give the correct result. For example:

```javascript
let bytesArray = [];
for (let i = 0; i < 256; i++) {
  bytesArray.push(i);
}

let b1 = Buffer.from(bytesArray);
console.log(b1.length); // prints 256

mystring = b1.toString('utf-8');
b2 = Buffer.from(mystring, 'utf-8');
console.log(b2.length); // prints 512
```

sdumitriu replied:
That is "expected". The error is not in `b2 = Buffer(...)`, but in `b1.toString('utf-8')`. `toString(encoding)` assumes that the contents of the buffer **is** in the specified encoding, but the bytes between 128 and 255 are **not** valid UTF-8 bytes. `mystring` ends with a bunch of `�` (U+FFFD), which, when turned back into bytes, takes 3 bytes. So b2 contains 128 valid bytes for the first valid 128 characters, and 3x128 for the U+FFFD invalid characters.

If you change your code to go only up to 127, both buffers will print 128 as the length of the buffer.

sdumitriu commented on Utilities/Administration/Backup-To-JSON/restore_json_backup.py:
```diff
@@ -0,0 +1,209 @@
+    return None
+  with open(subjectJsonPath, 'r') as f:
+    subjectData = json.loads(f.read())
+  if subjectData["@path"] == subjectPath:
+    return subjectData
```
Is this check needed? Why don't we do the same check for forms?

IntegralProgrammer replied:
The check is not needed in typical use cases, however, with subject names not assigned by UUID, it guards against the wrong subject being selected. For example, `/Subjects/abcd` and `/Subjects/abcd/abcd` are both backed up as `abcd.json` yet would contain different `@path` values. This is not an issue for Forms as all `cards:Form`s are direct descendants of the `/Forms/` JCR node - unlike `cards:Subject`s, there are no `cards:Form`s that are children of `cards:Form`s.

OrdiNeu commented on Utilities/Administration/Backup-To-JSON/restore_json_backup.py:
```diff
@@ -0,0 +1,209 @@
+if "CARDS_URL" in os.environ:
+  CARDS_URL = os.environ["CARDS_URL"].rstrip('/')
+
+ADMIN_PASSWORD = "admin"
```
This should be documented in the `README.md` file

IntegralProgrammer replied:
Fixed in commits https://github.com/data-team-uhn/cards/pull/1040/commits/2719176948c71b6e4844fe1655200bf1062ad16b and https://github.com/data-team-uhn/cards/pull/1040/commits/66c1d21827c54925b35b9ad68f40b471d0a24368. :heavy_check_mark:

OrdiNeu commented on Utilities/Administration/Backup-To-JSON/restore_json_backup.py:
```diff
@@ -0,0 +1,209 @@
+  print("Usage: python3 restore_json_backup.py BACKUP_DIRECTORY FORM_LIST_FILE SUBJECT_LIST_FILE")
+  sys.exit(1)
+
+BACKUP_DIRECTORY = sys.argv[1]
```
Prefer [argparse](https://docs.python.org/3/library/argparse.html) to doing it like this, as it covers the functionality from lines 29:35.

IntegralProgrammer replied:
Fixed in commit https://github.com/data-team-uhn/cards/pull/1040/commits/a5ceef194fb74c2c44ca94772f3ffe6e6d293c0c. :heavy_check_mark:

OrdiNeu commented on Utilities/Administration/Backup-To-JSON/restore_json_backup.py:
```diff
@@ -0,0 +1,209 @@
+SUBJECT_LIST_FILE = sys.argv[3]
+
+CARDS_URL = "http://localhost:8080"
+if "CARDS_URL" in os.environ:
```
Why `os.environ` over an optional command-line argument?

IntegralProgrammer replied:
This is the way that it's done in the other Python utility programs for CARDS and the reason for that is so that the user can run `export CARDS_URL=...` and `export ADMIN_PASSWORD=...` at the beginning of their shell and the subsequent calls to the various Python utility programs will act on the specified CARDS server.

marta- commented on Utilities/Administration/Backup-To-JSON/backup_recorder.js:
```diff
@@ -0,0 +1,262 @@
+      simplifiedAnswer[key] = answer[key];
+    }
+  }
+  if (answer["jcr:primaryType"] === "cards:FileAnswer") {
```
Do we need something similar for DicomAnswer?

IntegralProgrammer replied:
We would, and it would be the same logic as for what's happening here with `cards:FileAnswer` but `cards:DicomAnswer` is not yet present in this `CARDS-1759` branch so we can either rebase this branch with the latest `dev` or merge as it is and create a new task to add backup support for `cards:DicomAnswer` - which would probably simply consist of replacing `answer["jcr:primaryType"] === "cards:FileAnswer"` with `answer["jcr:primaryType"] === "cards:FileAnswer" || answer["jcr:primaryType"] === "cards:DicomAnswer"`.

marta- replied:
Rebase, since dicom is already in dev so technically this code is a bit stale.

IntegralProgrammer replied:
Rebased and `cards:DicomAnswer` handling code added in commit https://github.com/data-team-uhn/cards/pull/1040/commits/08620960b473d63e5b0530ddb2803a87643dc35a. :heavy_check_mark:

marta- commented on Utilities/Administration/Backup-To-JSON/backup_recorder.js:
```diff
@@ -0,0 +1,262 @@
+  return simplifiedAnswer;
+};
+
+const flattenAnswerSection = (answerSection) => {
+  let answers = [];
+  for (let key in answerSection) {
+    if (typeof(answerSection[key]) !== "object") {
+      continue;
+    }
+    if (!("sling:resourceSuperType" in answerSection[key])) {
+      continue;
+    }
+    if (answerSection[key]["sling:resourceSuperType"] === "cards/Answer") {
+      answers.push(answerSection[key]);
+    }
+    if (answerSection[key]["sling:resourceType"] === "cards/AnswerSection") {
+      answers = answers.concat(flattenAnswerSection(answerSection[key]));
+    }
+  }
+  return answers;
+};
```
Hopefully this can be replaced by the new processor in #1052 .

IntegralProgrammer replied:
Similar to https://github.com/data-team-uhn/cards/pull/1040#discussion_r900522173 should we merge this PR as is and then create a new task to switch to the processor implemented in https://github.com/data-team-uhn/cards/pull/1052?

marta- replied:
For this one we can merge as is and refactor after, maybe at the same time as the `cleanup` processor mentioned in the comment below.

marta- commented on Utilities/Administration/Backup-To-JSON/backup_recorder.js:
```diff
@@ -0,0 +1,262 @@
+  return answers;
+};
+
+const cleanupForm = (formObject) => {
```
Should we replace this with a json processor?

IntegralProgrammer replied:
Maybe - but implementing the new JSON processor would be a new task and removing this code from `backup_recorder.js` would be another new task, right?

marta- replied:
Yes separate task.

OrdiNeu commented on Utilities/Administration/Backup-To-JSON/restore_json_backup.py:
```diff
@@ -0,0 +1,209 @@
+    thisSectionQPath = '/'.join(['', 'Questionnaires', questionnaireNode] + questionnaireSectionNodes[0:i+1])
+    print("Creating answer section at {} for question section {}".format(thisSectionFPath, thisSectionQPath))
+    createAnswerSectionInJcr(thisSectionFPath, thisSectionQPath)
+
```
Is there any interest in saving Questionnaires as well? Since they can be manually added/removed, and backing up Forms for a custom questionnaire and then loading it causes problems

IntegralProgrammer replied:
Yes - this is a very good idea but should be implemented in a new task.

OrdiNeu commented on modules/webhook-backup/src/main/java/io/uhndata/cards/webhookbackup/WebhookBackupTask.java:
```diff
@@ -0,0 +1,284 @@
+    {
+        String taskUpdateMessage = emojii + " Backup " + status + " for data modified ";
+        if (upper == null) {
+            taskUpdateMessage += "after " + lower.format(DateTimeFormatter.ofPattern(DATE_TIME_JCR_FORMAT));
```
It appears possible to reach a case where `lower` here is `null`, causing an NRE (trigger an Endpoint-based backup with neither lower nor upper bound.

IntegralProgrammer replied:
That case is actually unreachable. Looking at the following code ...

https://github.com/data-team-uhn/cards/blob/66c1d21827c54925b35b9ad68f40b471d0a24368/modules/webhook-backup/src/main/java/io/uhndata/cards/webhookbackup/WebhookBackupEndpoint.java#L63-L72

...if `dateLowerBound` and `dateUpperBound` are both `null`, the webhook backup task will run in `manualToday` mode and thus will automatically obtain a lower bound of the start of today and an upper bound of the start of tomorrow.

OrdiNeu replied:
Oh, I see. I saw the `exportLowerBound` being set to `null` and never changed, but didn't notice the `manualToday` mode's call to `doNightlyExport()` :heavy_check_mark: 
https://github.com/data-team-uhn/cards/blob/66c1d21827c54925b35b9ad68f40b471d0a24368/modules/webhook-backup/src/main/java/io/uhndata/cards/webhookbackup/WebhookBackupTask.java#L56-L62

OrdiNeu commented on modules/webhook-backup/src/main/java/io/uhndata/cards/webhookbackup/WebhookBackupTask.java:
```diff
@@ -0,0 +1,284 @@
+
+    private void postToSlack(String msg)
+    {
+        final String slackNotificationsUrl = System.getenv("SLACK_NOTIFICATIONS_URL");
```
Do we have documentation somewhere for what this can be set to? I don't see a reference to this elsewhere in the codebase.

IntegralProgrammer replied:
I renamed this environment variable to `SLACK_BACKUP_NOTIFICATIONS_URL` in commits https://github.com/data-team-uhn/cards/pull/1040/commits/9623027ff1a9c6794c5b7c9a4b088f8dcbbc2c21 and https://github.com/data-team-uhn/cards/pull/1040/commits/752956d9ed7c35927ed4bb2254bb8fbbeb272e9d so it's slightly more intuitive.

In a new task, let's create a Markdown document listing all the applicable environment variables and what they do. @marta-

OrdiNeu commented on modules/webhook-backup/src/main/java/io/uhndata/cards/webhookbackup/WebhookBackupTask.java:
```diff
@@ -0,0 +1,284 @@
+
+    private void output(String input, String filename) throws IOException
+    {
+        final String backupWebhookUrl = System.getenv("BACKUP_WEBHOOK_URL");
```
Might be worth it to be safe and throw an error in the log if this environment variable is not set, since people triggering this might not know to set it.

IntegralProgrammer replied:
Implemented in commit https://github.com/data-team-uhn/cards/pull/1040/commits/7ebb3eff42ca2db27a79631b557282aa91f889a9 :heavy_check_mark:

OrdiNeu reviewed (changes requested):
Works well, but needs a bit more documentation, as `WebhookBackupTask.java` and `restore_json_backup.py` are a bit hard to read as-is.

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Add support for cards:DicomAnswer answers

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Add documentation to README.md describing the use of the CARDS_URL and ADMIN_PASSWORD environment variables

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

In restore_json_backup.py, switch from using sys.argv to argparse

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Rename environment variable SLACK_NOTIFICATIONS_URL to SLACK_BACKUP_NOTIFICATIONS_URL

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Handle the case where the BACKUP_WEBHOOK_URL environment variable is not defined

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Fix line-length checkstyle violation

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Fix README.md markdown syntax

OrdiNeu reviewed (changes requested):
Bug found:
1. Rebase the branch on `dev`
2. `mvn clean install` as normal
3.  `export BACKUP_WEBHOOK_URL="http://localhost:8012"` and run in cards4proms mode (`./start_cards.sh --dev -P cards4proms`)
4. Make a Visit Information form with either clinic (I chose 6012) and a date a few days in the future (for me, June 25th 2022)
5. Run the backup_recorder as normal (`nodejs backup_recorder.js /home/fnguyen/Documents/tmp/backup/`)
6. Run the exporter (`http://localhost:8080/Subjects.webhookbackup?dateLowerBound=2022-06-23T00:00:00`)
7. Only some of the forms will be pushed, and none of the subjects will be. In the error log, one can see that there was an error 
while serializing one of the forms (the Visit Information form)

Error message:
```
 23.06.2022 15:10:36.118 *ERROR* [Thread-212] org.apache.sling.extensions.threaddump.internal.Activator Uncaught exception in Thread Thread[Thread-212,5,main]
 java.lang.NullPointerException: null 
     at io.uhndata.cards.dataentry.internal.serialize.labels.AbstractResourceLabelProcessor.getLabelForResource(AbstractResourceLabelProcessor.java:72) [io.uhndata.cards.dataentry:0.9.0.SNAPSHOT]
     at io.uhndata.cards.dataentry.internal.serialize.labels.ResourceLabelProcessor.getAnswerLabel(ResourceLabelProcessor.java:88) [io.uhndata.cards.dataentry:0.9.0.SNAPSHOT]
     at io.uhndata.cards.dataentry.internal.serialize.labels.SimpleAnswerLabelProcessor.addProperty(SimpleAnswerLabelProcessor.java:93) [io.uhndata.cards.dataentry:0.9.0.SNAPSHOT]
     at io.uhndata.cards.dataentry.internal.serialize.labels.ResourceLabelProcessor.leave(ResourceLabelProcessor.java:64) [io.uhndata.cards.dataentry:0.9.0.SNAPSHOT]
     at io.uhndata.cards.serialize.ResourceToJsonAdapterFactory.lambda$leaveNode$6(ResourceToJsonAdapterFactory.java:219) [io.uhndata.cards.utils:0.9.0.SNAPSHOT]
     at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
     at io.uhndata.cards.serialize.ResourceToJsonAdapterFactory.leaveNode(ResourceToJsonAdapterFactory.java:219) [io.uhndata.cards.utils:0.9.0.SNAPSHOT]
     at io.uhndata.cards.serialize.ResourceToJsonAdapterFactory.serializeNode(ResourceToJsonAdapterFactory.java:122) [io.uhndata.cards.utils:0.9.0.SNAPSHOT]
     at io.uhndata.cards.serialize.ResourceToJsonAdapterFactory.lambda$processChildren$4(ResourceToJsonAdapterFactory.java:201) [io.uhndata.cards.utils:0.9.0.SNAPSHOT]
     at io.uhndata.cards.serialize.internal.DeepProcessor.processChild(DeepProcessor.java:60) [io.uhndata.cards.utils:0.9.0.SNAPSHOT]
     at io.uhndata.cards.serialize.ResourceToJsonAdapterFactory.processChildren(ResourceToJsonAdapterFactory.java:201) [io.uhndata.cards.utils:0.9.0.SNAPSHOT]
     at io.uhndata.cards.serialize.ResourceToJsonAdapterFactory.serializeNode(ResourceToJsonAdapterFactory.java:121) [io.uhndata.cards.utils:0.9.0.SNAPSHOT]
     at io.uhndata.cards.serialize.ResourceToJsonAdapterFactory.getAdapter(ResourceToJsonAdapterFactory.java:92) [io.uhndata.cards.utils:0.9.0.SNAPSHOT]
     at org.apache.sling.adapter.internal.AdapterManagerImpl.getAdapter(AdapterManagerImpl.java:121) [org.apache.sling.adapter:2.2.0]
     at org.apache.sling.api.adapter.SlingAdaptable.adaptTo(SlingAdaptable.java:104) [org.apache.sling.api:2.25.0]
     at org.apache.sling.jcr.resource.internal.helper.jcr.JcrNodeResource.adaptTo(JcrNodeResource.java:154) [org.apache.sling.jcr.resource:3.2.0]
     at io.uhndata.cards.webhookbackup.WebhookBackupTask.getFormAsJson(WebhookBackupTask.java:210) [io.uhndata.cards.modules-webhook-backup:0.9.0.SNAPSHOT]
     at io.uhndata.cards.webhookbackup.WebhookBackupTask.doManualExport(WebhookBackupTask.java:111) [io.uhndata.cards.modules-webhook-backup:0.9.0.SNAPSHOT]
     at io.uhndata.cards.webhookbackup.WebhookBackupTask.run(WebhookBackupTask.java:80) [io.uhndata.cards.modules-webhook-backup:0.9.0.SNAPSHOT]
     at java.base/java.lang.Thread.run(Thread.java:829)
```

The partial backup will not be capable of loading, since none of the Subjects are able to backup. This bug appears to be related to the new Resource question type introduced in #1043

OrdiNeu committed:
CARDS-1784 / CARDS-1759 - Fix missing type hint

marta- commented:
Rebased on latest `dev` with the fix for exporting & importing resource questions (export fix in `dev`, import fix in 38eee4c).

sdumitriu commented on modules/webhook-backup/src/main/java/io/uhndata/cards/webhookbackup/WebhookBackupTask.java:
```diff
@@ -0,0 +1,290 @@
+        }
+    }
+
+    private void postToSlack(String msg)
```
I don't really like that this class is also responsible for talking to Slack.

IntegralProgrammer replied:
There is almost an identical method over in `io.uhndata.cards.proms.slacknotifications`

https://github.com/data-team-uhn/cards/blob/c122630baabf626cbdfde43914b9a83db21449c2/proms-resources/backend/src/main/java/io/uhndata/cards/proms/slacknotifications/SlackNotificationsTask.java#L71-L81

Maybe what we should do is, in a future task, move the `postToSlack` method to its own commons module so that it can be called both from here as well as from `io.uhndata.cards.proms.slacknotifications.run`?

sdumitriu replied:
Sounds good.

sdumitriu commented on modules/webhook-backup/src/main/java/io/uhndata/cards/webhookbackup/WebhookBackupTask.java:
```diff
@@ -0,0 +1,290 @@
+    {
+        final String slackNotificationsUrl = System.getenv("SLACK_BACKUP_NOTIFICATIONS_URL");
+        if (slackNotificationsUrl == null) {
+            LOGGER.warn(
```
This shouldn't be a warning, the notification is not a critical part of the backup process.

IntegralProgrammer replied:
Slack notification is not a _critical_ part of the backup process but I still consider it very important. The reason for that is because, by default, each day of data entry is backed up once and only once. Should one of those days fail to backup, it's important that this be known to the administrators as the backed-up data would be incomplete. Consider the following example:

- `/Forms/1234` is created on June 1st 2022
- Backup successfully runs and `/Forms/1234` is saved
- `/Forms/5678` is created on June 2nd 2022
- Backup runs but fails. `/Forms/5678` is _NOT_ saved
- `/Forms/9012` is created on June 3rd 2022
- Backup successfully runs and `/Forms/9012` is saved.
- At this point, the backup contains `/Forms/1234` and `/Forms/9012` but `/Forms/5678` is missing due to the failed backup. Administrators should know that this problem occurred.

Perhaps something else that should be added to help verify that backups are complete is to save the value of `jcr:lastModified` associated with each _Form_ during the export of the list of all _Forms_ stored by a CARDS instance. That way, when given a list of _Forms_, and their associated `jcr:lastModified` values, a verification program can check that a JSON file with a matching value of `jcr:lastModified` can be found for each entry in the Form list backup.

IntegralProgrammer replied:
> Perhaps something else that should be added to help verify that backups are complete is to save the value of `jcr:lastModified` associated with each _Form_ during the export of the list of all _Forms_ stored by a CARDS instance. That way, when given a list of _Forms_, and their associated `jcr:lastModified` values, a verification program can check that a JSON file with a matching value of `jcr:lastModified` can be found for each entry in the Form list backup.

Added in https://github.com/data-team-uhn/cards/pull/1040/commits/91b763140b71e24b0234ad46890583815dda015a :heavy_check_mark:

sdumitriu commented on modules/webhook-backup/src/main/java/io/uhndata/cards/webhookbackup/WebhookBackupTask.java:
```diff
@@ -0,0 +1,290 @@
+        try (ResourceResolver resolver = this.resolverFactory.getServiceResourceResolver(null)) {
+            Set<String> changedForms = new HashSet<>();
+            String query = String.format(
+                "SELECT * FROM [" + cardsType + "] AS form"
```
It's not always a `form`, so a more generic name for the selector should be used.

IntegralProgrammer replied:
Fixed in https://github.com/data-team-uhn/cards/pull/1040/commits/92756590bab20c4c16808d0990203ea8aa27207c :heavy_check_mark:.

sdumitriu commented on modules/webhook-backup/src/main/java/io/uhndata/cards/webhookbackup/WebhookBackupTask.java:
```diff
@@ -0,0 +1,290 @@
+
+    private String getFormAsJson(String formPath) throws IOException
+    {
+        String formDataUrl = String.format("%s.data.deep", formPath);
```
`data` only applies to subjects and questionnaires.

```suggestion
        String formDataUrl = String.format("%s.deep", formPath);
```

sdumitriu commented on Utilities/Administration/Backup-To-JSON/backup_recorder.js:
```diff
@@ -0,0 +1,268 @@
+  let formFileName = path.basename(formName);
+  if (validateFormFileName(formFileName)) {
+    fs.writeFileSync(
+      BACKUP_DIRECTORY + "/Forms/" + formFileName + ".json",
```
This shouldn't be publicly accessible, but I wonder if some sanitizing of the input is possible (or even needed) to avoid writing to `/backup/dir/Forms/../../../etc/passwd`.

IntegralProgrammer replied:
Adding an additional step here to sanitize the input is not necessary as it is already validated by the `validateFormFileName` method.

https://github.com/data-team-uhn/cards/blob/38eee4c26ce1bb739e01a79adb2da3ea391dffa1/Utilities/Administration/Backup-To-JSON/backup_recorder.js#L169-L180

Any input that contains characters outside of the `allowedChars` set will test as invalid - therefore `../../../etc/passwd` would fail the `validateFormFileName` test.

sdumitriu replied:
Oh, that's exactly what I wanted, somehow I missed the call to `validateFormFileName` right above my comment.

sdumitriu reviewed (commented):
Should the backup always be enabled? I was expecting it to be opt-in. Right now, the module is always loaded, the backup job is enabled, and not defining the two required environment variables is considered an exception.

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Address PR review comment - It's not always a form, so a more generic name for the selector should be used

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

Address PR review comment - `data` only applies to subjects and questionnaires

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

IntegralProgrammer committed:
CARDS-1759: Create a nightly data export task that will serialize all subjects that were modified that day and POST that data to a backup service webhook

When creating the FormListBackup and SubjectListBackup files, include the jcr:lastModified date/time for each Form or Subject so that an administrative utility can read through a FormListBackup file and check that each linked file under Forms/ has a matching jcr:lastModified date/time. This way, administrators can be sure that their backup reflects the state of CARDS at the time of backup without any missing data that may have resulted from a failed nightly backup

sdumitriu commented on Utilities/Administration/Backup-To-JSON/backup_recorder.js:
```diff
@@ -0,0 +1,272 @@
+  for (let i = 0; i < jsonValue.length; i++) {
+    bytesArray.push(jsonValue.charCodeAt(i));
+  }
+  return Buffer.from(bytesArray);
```
Actually this is wrong, non-ASCII characters will be mangled by this. From the documentation of `Buffer.from`:

> Allocates a new Buffer using an array of bytes in the range 0 – 255. Array entries outside that range will be truncated to fit into it.

Test with this string in a free text input: `aă-sș—tț`

IntegralProgrammer replied:
I tested this by writing your example string (`aă-sș—tț`) to a file, uploading it to a Form based on the _DICOM Imaging Test_ Questionnaire (in the CARDS `--test` mode) and then triggering a backup. With an unmodified `backup_recorder.js`, the data was saved properly in the `blobs` directory. However, when I modified `backup_recorder.js` to use `return Buffer.from(jsonValue, 'utf-8')` for the `getFileDataBuffer` method, data was mangled when saved in the `blobs` directory.

Do you have a test case where non-ASCII characters are being mangled upon backup?

sdumitriu replied:
OK, this doesn't make sense, but turns out it's a chain reaction of mis-decoding and mis-encoding of characters that in the end cancels out.

First, this is only used for opaque files used in answers, not for the actual JSON itself.

Since the file is an opaque blob, an encoding for it doesn't make sense, so it is not UTF-8 or anything else, it's a byte stream. It *could* be an UTF-8 text, but it could also be an ISO-8859-1 or just a binary.

Since we need to embed it somehow inside a JSON, which is text, it must somehow be turned into a piece of text, and we do that by reading it as an ISO-8859-1 string, which is a single byte encoding. As a Java String, it is internally in UTF-16, but putting it in the JSON and serializing into the HTTP request body turns it into UTF-8.

So far we have binary -> ISO-8859-1 -> UTF-8; that's what the backup application receives. During this chain of transformations, all the characters are in the 0-255 code range, which is the same in both Unicode and ISO.

Then on the node side, the UTF-8 JSON is turned, character by character, into bytes, which is then serialized as a binary blob into a file.

So, the string isn't really UTF-8, it is a binary file masquerading as UTF-8 characters.
